### PR TITLE
Fix Grounding DINO text position embeddings being truncated to int64

### LINKED ISSUE #47674: Grounding DINO text position embeddings are truncated to int64, collapsing 93% of them to zero
### System Info

```
- transformers version: 5.15.0.dev0 (main @ 1ce2865a97)
- Platform: Windows-11-10.0.26200-SP0
- Python version: 3.12.0
- PyTorch version: 2.13.0+cpu
- Accelerate version: 1.14.0
```

### Who can help?

@yonigozlan @molbap

### Information

- [X] The official example scripts
- [X] My own modified scripts

### Tasks

- [X] My own task or dataset (give details below)

### Reproduction

`GroundingDinoEncoderLayer.get_text_position_embeddings` builds the text position embeddings from `text_position_ids`, which is an int64 tensor:

```python
if text_position_ids is not None:
    text_position_embedding = encode_sinusoidal_position_embedding(
        text_position_ids[..., None], num_pos_feats=self.d_model
    )
```

`encode_sinusoidal_position_embedding` ends with `return torch.cat(embeddings, dim=-1).to(pos_tensor.dtype)`. With an int64 `pos_tensor` that truncates every sin/cos value in `[-1, 1]` toward zero, so the whole embedding collapses to `{-1, 0, 1}`:

```python
import torch
from transformers.models.grounding_dino.modeling_grounding_dino import encode_sinusoidal_position_embedding

pos_ids = torch.arange(8).unsqueeze(0)          # int64, as the model builds it
emb_int = encode_sinusoidal_position_embedding(pos_ids[..., None], num_pos_feats=256)
emb_flt = encode_sinusoidal_position_embedding(pos_ids.float()[..., None], num_pos_feats=256)

print(emb_int.dtype, sorted(set(emb_int.flatten().tolist())))
print(emb_flt.dtype, float(emb_flt.min()), float(emb_flt.max()))
print("collapsed to zero:", (emb_int == 0).float().mean().item())
```

```
torch.int64 [-1, 0, 1]
torch.float32 -0.9999997615814209 1.0
collapsed to zero: 0.934
```

93.4% of the text position embedding becomes exactly zero at the real `d_model=256`.

This is not a rarely-reached branch. `GroundingDinoModel.forward` builds the ids at `modeling_grounding_dino.py:2062` via `generate_masks_with_special_tokens_and_transfer_map`, which returns `torch.clamp(position_ids, min=0).to(torch.long)` (`:1854`), and passes them at `:2168` with `text_position_embedding=None`. So the `text_position_ids is not None` branch is taken on every forward pass, and the other branch is dead on the model path. Intercepting the method during a normal `GroundingDinoForObjectDetection` forward confirms it: `text_position_ids` is `torch.int64`, and so is the returned embedding.

### Regression source

Commit `c9de1097ee` (#41693, "Refactor ViT to updated standards", 8 May 2026) replaced the model's own `get_sine_pos_embed` with the shared `encode_sinusoidal_position_embedding`. The old helper ended with

```python
position_embeddings = torch.cat(position_embeddings, dim=-1)
return position_embeddings
```

with no cast, so it returned float32 whatever the input dtype was. The new one casts back to the input dtype, which is a no-op for the other callers in the codebase (`conditional_detr`, `dab_detr`, `lw_detr`, `rf_detr` all pass float anchor coordinates) but truncates for this one, which passes token position ids.

That commit changed `modeling_grounding_dino.py` and `modeling_mm_grounding_dino.py` and touched no test file, so the checked-in integration expectations still describe the pre-refactor behavior. Running the `@slow` integration tests against the real `IDEA-Research/grounding-dino-tiny` checkpoint on current main:

```
tests/models/grounding_dino/test_modeling_grounding_dino.py::GroundingDinoModelIntegrationTests
3 failed, 1 skipped

test_cross_attention_mask            Greatest absolute difference: 0.5683 (up to 0.001 allowed)
test_inference_object_detection_head Greatest absolute difference: 0.3003 (up to 0.001 allowed)
test_grounding_dino_loss             failed
```

`mm_grounding_dino` inherits the same encoder layer and has the same defect at `modeling_mm_grounding_dino.py:1010`.

### Expected behavior

The position ids should be cast to the text feature dtype before the sinusoid is computed, so the embedding keeps its values, and stays consistent with the model dtype in half precision:

```python
text_position_ids[..., None].to(text_features.dtype)
```

With that change, `test_cross_attention_mask` passes and the object detection head divergence halves (0.3003 to 0.1401). The remaining divergence suggests the same refactor changed at least one more thing that I have not identified; I would like to open a PR for the part that is established, and flag the rest.

### PR DESCRIPTION
**Pull request by Yigtwxx**
[CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=47675)

# What does this PR do?

`GroundingDinoEncoderLayer.get_text_position_embeddings` passes the int64 `text_position_ids` straight into `encode_sinusoidal_position_embedding`, which ends with `.to(pos_tensor.dtype)`. Every sin/cos value in `[-1, 1]` is therefore truncated toward zero and the embedding collapses to `{-1, 0, 1}` — at the real `d_model=256`, 93.4% of it becomes exactly zero.

The branch is taken on every forward: `GroundingDinoModel.forward` calls the encoder with `text_position_embedding=None` and `text_position_ids=position_ids` (`modeling_grounding_dino.py:2166-2168`), where `position_ids` comes from `generate_masks_with_special_tokens_and_transfer_map` and is `.to(torch.long)` (`:1854`). The other branch, which does `.float()` before calling, is dead on the model path — and it is the one that shows the intended dtype.

This casts the ids to the text feature dtype instead, which keeps the values and also keeps the embedding in the model dtype under half precision:

```python
-    text_position_ids[..., None], num_pos_feats=self.d_model
+    text_position_ids[..., None].to(text_features.dtype), num_pos_feats=self.d_model
```

`mm_grounding_dino` inherits this encoder layer through modular, so the change propagates to its generated file without a separate edit.

## Where it came from

`c9de1097ee` (#41693) replaced the model's own `get_sine_pos_embed` with the shared `encode_sinusoidal_position_embedding`. The old helper returned `torch.cat(position_embeddings, dim=-1)` with no cast, so it produced float32 regardless of input dtype. The new one casts back to the input dtype, which is a no-op for every other caller (`conditional_detr`, `dab_detr`, `lw_detr`, `rf_detr` pass float anchor coordinates) and destructive for this one, which passes token position ids.

That commit changed the two modeling files and no test file, so the checked-in integration expectations still describe the pre-refactor behavior.

## Tests

`@slow` integration tests against the real `IDEA-Research/grounding-dino-tiny` checkpoint. These ran on CPU, so they compare against the `(None, None)` entry of the `Expectations` blocks; I do not have the A10 the slow CI uses, so the exact deltas there may differ:

| | main | this PR |
|---|---|---|
| `GroundingDinoModelIntegrationTests` | 3 failed, 1 skipped | 2 failed, **1 passed**, 1 skipped |
| `test_cross_attention_mask` | failed, max abs diff 0.5683 | **passes** |
| `test_inference_object_detection_head` | failed, max abs diff 0.3003 | failed, max abs diff **0.1401** |
| `test_grounding_dino_loss` | failed | failed |

So one of the repo's own integration expectations goes green and another's divergence halves, without those expectations being touched.

I want to be precise about what this does not do: it does not fully restore them. `test_grounding_dino_loss` still fails and the detection head is still outside tolerance, which suggests the same refactor changed at least one more thing. I could not identify it, so I have left it rather than guess, and I would rather land the part that is established than bundle it with something unverified. Happy to keep digging if that is preferred.

Fast tests, both models, unchanged:

```
tests/models/grounding_dino tests/models/mm_grounding_dino
main    191 passed, 323 skipped
branch  191 passed, 323 skipped
```

`ruff check`, `ruff format --check`, `modular_conversion` and `copies` are clean.

Note this changes Grounding DINO's numerical outputs, in the direction of the expectations that predate the refactor. Let me know if you would like the title marked accordingly.

Fixes #47674

Disclosure: this fix was implemented with the assistance of a code agent, and reviewed and validated by me.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://huggingface.co/docs/transformers/contributing) and the
      [Pull Request](https://huggingface.co/docs/transformers/pr_checks) checks?
- [x] Was this discussed/approved via a Github issue? https://github.com/huggingface/transformers/issues/47674
- [ ] Did you make sure to update the documentation with your changes according to the [guidelines](https://github.com/huggingface/transformers/tree/main/docs)?
- [ ] Did you write any new necessary [tests](https://huggingface.co/docs/transformers/testing)?

## Who can review?

@yonigozlan @molbap

### CODE REVIEW

**Comment by molbap** [comment]
I think the cast should be at `encode_sinusoidal_position_embedding`, or  tensors be created with the right dtype from init!
